### ISSUE TITLE
randint, compile and dynamic shape

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#122405]
### 🐛 Describe the bug

The following call to randint fails when the shape of the function input changes:

```python
import torch

@torch.compile()
def get_traj_idx(lengths: torch.Tensor, num_slices: int) -> torch.Tensor:
    return torch.randint(lengths.shape[0], (num_slices,), device=lengths.device)

lengths = torch.zeros(10, dtype=torch.long)
get_traj_idx(lengths, num_slices=4)
lengths = torch.zeros(11, dtype=torch.long)
get_traj_idx(lengths, num_slices=4)
```

Error log:
```
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: prims::inductor_randint() Expected a value of type 'int' for argument 'high' but instead found type 'Node'.
Position: 1
Value: arg0_1
Declaration: prims::inductor_randint(SymInt low, SymInt high, SymInt[] size, Tensor seed) -> Tensor
Cast error details: Unable to cast Python instance of type <class 'torch.fx.node.Node'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

```

### Versions

Nightly

cc @pbelevich @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

**Comment by ezyang** [comment]
oh bruh, I also ran into this at https://github.com/pytorch/pytorch/pull/122375

I pinged @jansel about this because I'm not sure how this code ever worked at all

**Comment by jansel** [comment]
```diff
diff --git a/torch/_inductor/fx_passes/replace_random.py b/torch/_inductor/fx_passes/replace_random.py
index 45863aac16e..59d4c389122 100644
--- a/torch/_inductor/fx_passes/replace_random.py
+++ b/torch/_inductor/fx_passes/replace_random.py
@@ -131,9 +131,9 @@ def replace_randint(
     layout=None,
     pin_memory=None,
 ):
-    def replacement(size):
+    def replacement(low, high, size):
         result = inductor_prims.randint(low, high, size, inductor_prims.seed(device))
         return result.to(dtype)
 
     device = get_device(device)
-    match.replace_by_example(replacement, [size])
+    match.replace_by_example(replacement, [low, high, size])
diff --git a/torch/_inductor/lowering.py b/torch/_inductor/lowering.py
index f44c34d0d78..51dfe99c67b 100644
--- a/torch/_inductor/lowering.py
+++ b/torch/_inductor/lowering.py
@@ -2096,8 +2096,8 @@ def inductor_randint(
         return ops.randint64(
             seed_loader([]),
             ops.index_expr(random_pos(index), torch.int32),
-            low,
-            high,
+            ops.index_expr(low, torch.int64),
+            ops.index_expr(high, torch.int64),
         )
 
     return Pointwise.create(
```

The above patch results in a new error:
```
  File "/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/sympy/printing/printer.py", line 331, in _print
    return printmethod(expr, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1192, in _print_Symbol
    assert self.symbol_to_source.get(expr), (
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError: s1 (could be from ['input0']) not in {s0: ["L['t0']"], s1: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

cc @ezyang

**Comment by jansel** [comment]
It works if I comment out:
```diff
diff --git a/torch/fx/experimental/symbolic_shapes.py b/torch/fx/experimental/symbolic_shapes.py
index ebf0a1c9a2f..0ff4726b068 100644
--- a/torch/fx/experimental/symbolic_shapes.py
+++ b/torch/fx/experimental/symbolic_shapes.py
@@ -3231,7 +3231,7 @@ class ShapeEnv:
         for guard in self.guards:
             if self._maybe_evaluate_static(guard.expr) is not None:
                 continue
-            issue_guard(guard)
+            # issue_guard(guard)
 
         # 3. Every symbol must be within its value range (this handles 0/1
         # specialization too).
```

So that patch should fix things if we can get the guard printing sorted out.

**Comment by vmoens** [comment]
This probably doesn't help but this work!
```python
import torch

@torch.compile()
def get_traj_idx(lengths: torch.Tensor, num_slices: int) -> torch.Tensor:
    # return torch.randint(lengths.shape[0], (num_slices,), device=lengths.device)
    return (torch.rand((num_slices,), device=lengths.device) * lengths.shape[0]).floor()

lengths = torch.zeros(10, dtype=torch.long)
get_traj_idx(lengths, num_slices=4)
lengths = torch.zeros(11, dtype=torch.long)
get_traj_idx(lengths, num_slices=4)
```

**Comment by jansel** [comment]
@vmoens if you are looking for a workaround a better one might be:
```
torch._inductor.config.fallback_random = True
```
or 
```
torch.compile(options={"fallback_random": True})
```

**Comment by jansel** [comment]
Should be fixed by #122428
